sql/catalyst/src/main/scala/org/apache/spark/sql/types/UnionType.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.types
+
+import org.apache.spark.annotation.DeveloperApi
+
+@DeveloperApi
+case class UnionType(types: Seq[DataType]) extends DataType {
```

Comment:
> we can either create a table/test with a simpler schema

Yes, but isn't the goal here to test support for the thrift serdes? Writing a non-thrift type defeats that. And writing a different thrift type means you'd have to generate new data (`sql/hive/src/test/resources/data/files/complex.seq`). Not an insurmountable problem, but forking `Complex.java` from Hive 0.13 is not horrible either.